In [2]:
import transformers

#Set to avoid warning messages.
transformers.logging.set_verbosity_error()

### Usage of pipeline with a context

In [3]:
from transformers import pipeline

context="""
Earth is the third planet from the Sun and the only astronomical object
known to harbor life. While large volumes of water can be found
throughout the Solar System, only Earth sustains liquid surface water.
About 71% of Earth's surface is made up of the ocean, dwarfing
Earth's polar ice, lakes, and rivers. The remaining 29% of Earth's
surface is land, consisting of continents and islands.
Earth's surface layer is formed of several slowly moving tectonic plates,
interacting to produce mountain ranges, volcanoes, and earthquakes.
Earth's liquid outer core generates the magnetic field that shapes Earth's
magnetosphere, deflecting destructive solar winds.
"""

quan_pipeline = pipeline("question-answering",
                         model="deepset/minilm-uncased-squad2")

answer=quan_pipeline(question="How much of earth is land?",
             context=context)
print(answer)



/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/477 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/133M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/107 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

{'score': 0.9553400874137878, 'start': 323, 'end': 326, 'answer': '29%'}


In [4]:
print("\nAnother question :")
print(quan_pipeline( question="How are mountain ranges created?",
             context=context))


Another question :
{'score': 0.26154059171676636, 'start': 393, 'end': 465, 'answer': "Earth's surface layer is formed of several slowly moving tectonic plates"}


### Evaluation 1: Context 1


In [6]:
# Import the necessary module from the transformers library

from transformers import pipeline


# Define a string variable 'context' that holds the relevant information
# about Earth. This will be used by the question-answering model to find the answer.
context = """
Earth is the third planet from the Sun and the only astronomical object
known to harbor life. While large volumes of water can be found
throughout the Solar System, only Earth sustains liquid surface water.
About 71% of Earth's surface is made up of the ocean, dwarfing
Earth's polar ice, lakes, and rivers. The remaining 29% of Earth's
surface is land, consisting of continents and islands.
Earth's surface layer is formed of several slowly moving tectonic plates,
interacting to produce mountain ranges, volcanoes, and earthquakes.
Earth's liquid outer core generates the magnetic field that shapes Earth's
magnetosphere, deflecting destructive solar winds.
"""

# Create a question-answering pipeline using the 'deepset/minilm-uncased-squad2' model.
# This pipeline will handle both the question and the context to provide an answer.
quan_pipeline = pipeline("question-answering",
                         model="deepset/minilm-uncased-squad2")

# Use the pipeline to answer a specific question. Here, the question is
# "How much of earth is land?" and it will search for the answer in the
# provided 'context'.
answer = quan_pipeline(question="How much of earth is land?",
                       context=context)


# Print the answer extracted by the pipeline.
print(answer)

{'score': 0.9553400874137878, 'start': 327, 'end': 330, 'answer': '29%'}


### Evaluation 2: Context 2


In [8]:
# Import the necessary module from the transformers library
from transformers import pipeline

# Define a string variable 'context' that holds the relevant medical information.
context = """
Hypertension, also known as high blood pressure, is a common condition
in which the long-term force of the blood against your artery walls is
high enough that it may eventually cause health problems, such as heart disease.
Blood pressure is determined both by the amount of blood your heart pumps
and the amount of resistance to blood flow in your arteries. The more blood
your heart pumps and the narrower your arteries, the higher your blood pressure.
Hypertension is classified as primary (essential) hypertension, which has no identifiable cause,
and secondary hypertension, which is caused by an underlying condition such as
kidney disease or endocrine disorders. Managing hypertension typically involves lifestyle changes
and medications to lower blood pressure and reduce the risk of complications.
"""


# Create a question-answering pipeline using the 'deepset/minilm-uncased-squad2' model.
# This pipeline will handle both the question and the context to provide an answer.
quan_pipeline = pipeline("question-answering",
                         model="deepset/minilm-uncased-squad2")

# Use the pipeline to answer a specific question. Here, the question is
# "What is the most common type of hypertension?" and it will search for the answer in the
# provided 'context'.
answer = quan_pipeline(question="What is the most common type of hypertension?",
                       context=context)

# Print the answer extracted by the pipeline.
print(answer)

{'score': 0.26236692070961, 'start': 29, 'end': 48, 'answer': 'high blood pressure'}



#### Use of model: bert-large-uncased-whole-word-masking-finetuned-squad


In [10]:

# Create a question-answering pipeline using the 'bert-large-uncased-whole-word-masking-finetuned-squad' model
quan_pipeline = pipeline("question-answering",
                         model="bert-large-uncased-whole-word-masking-finetuned-squad")

# Ask the question
answer = quan_pipeline(question="What is the most common type of hypertension?",
                       context=context)

# Print the answer
print(answer)

{'score': 0.2682048976421356, 'start': 491, 'end': 523, 'answer': 'primary (essential) hypertension'}


#### Use of model: distilbert-base-uncased-distilled-squad

In [11]:

# Create a question-answering pipeline using the 'distilbert-base-uncased-distilled-squad' model
quan_pipeline = pipeline("question-answering",
                         model="distilbert-base-uncased-distilled-squad")

# Ask the question
answer = quan_pipeline(question="What is the most common type of hypertension?",
                       context=context)

# Print the answer
print(answer)

config.json:   0%|          | 0.00/451 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/265M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

{'score': 0.7166475653648376, 'start': 29, 'end': 48, 'answer': 'high blood pressure'}


#### Use of model: roberta-base-squad2

In [13]:

# Create a question-answering pipeline using the 'roberta-base-squad2' model
quan_pipeline = pipeline("question-answering",
                         model="deepset/roberta-base-squad2")

# Ask the question
answer = quan_pipeline(question="What is the most common type of hypertension?",
                       context=context)

# Print the answer
print(answer)

{'score': 0.005162023939192295, 'start': 29, 'end': 48, 'answer': 'high blood pressure'}


Observation:
  - accuracy is different for the same context with different models is different
  - Each model size is different, and some are big and others are light
  - model named: distilbert-base-uncased-distilled-squad, shows better performance with a score of 0.7166475653648376 with a correct answer.
  